numpy/core/src/common/ufunc_override.c

```diff
@@ -30,6 +32,11 @@ PyUFuncOverride_GetNonDefaultArrayUfunc(PyObject *obj)
     if (PyArray_CheckExact(obj)) {
         return NULL;
     }
+    /* Fast return for most common numpy scalar types */
+    if (PyObject_CheckExact_Type(obj, PyDoubleArrType_Type) || PyObject_CheckExact_Type(obj, PyIntArrType_Type)) {
+        return NULL;
+    }
+
```

Comment:
The alternative would be to put the special-casing inside `PyArray_LookupSpecial` (where there is already code for skipping basic python types).

Reply:
Putting the check inside of `PyArray_LookupSpecial` can be done and seems to be just as fast, but I decided not to at this moment. Considerations

* Pro: The fast path is also used for the other usage of `PyArray_LookupSpecial` (in `get_array_function` in `arrayfunction_override.c`)
* Small con: For the fast path `PyArray_LookupSpecial` would return NULL, and in `PyUFuncOverride_GetNonDefaultArrayUfunc` an unnecessary call to `PyErr_Occurred` is made. This does not seem 		expensive though
* Large con: The numpy scalars have `__array_function__` set, so `PyArray_LookupSpecial(obj, "__array_function__")` would return NULL, instead of the function. This is not an issue in the current codebase, `PyArray_LookupSpecial(obj, __array_function__)` does not seem to be called anywhere, but `PyArray_LookupSpecial_OnInstance(obj, __array_function__)` is called, and it would create an asymmetry between the two methods that is confusing.

@mhvk

Reply:
OK, that makes total sense - if the special method exists, it definitely should be returned!